### PR TITLE
Use the NS TTL Controller instead of cleanup pod

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @smarterclayton 
/cc @petr-muller @bbguimaraes @droslean

### CODE REVIEW

File: cmd/ci-operator/main.go [hunk 1]
```diff
@@ -599,110 +618,6 @@ func (o *options) initializeNamespace() error {
...
-							if [[ "${count}" -lt 1 ]]; then
-								count+=1
-								sleep ${WAIT} & wait
-								continue
-							fi
-							echo "No pods running for more than ${WAIT}s, deleting project ..."
-							oc delete project ${NAMESPACE}
-							exit 0
-						done
-						`,
```
Comment (bbguimaraes) [inline cmd/ci-operator/main.go]: RIP

**Comment by stevekuznetsov** [comment]
Needs https://github.com/openshift/release/pull/1141

File: cmd/ci-operator/main.go [hunk 2]
```diff
@@ -216,6 +213,7 @@ func bindOptions(flag *flag.FlagSet) *options {
 	flag.StringVar(&opt.namespace, "namespace", "", "Namespace to create builds into, defaults to build_id from JOB_SPEC. If the string '{id}' is in this value it will be replaced with the build input hash.")
 	flag.StringVar(&opt.baseNamespace, "base-namespace", "stable", "Namespace to read builds from, defaults to stable.")
 	flag.DurationVar(&opt.idleCleanupDuration, "delete-when-idle", opt.idleCleanupDuration, "If no pod is running for longer than this interval, delete the namespace. Set to zero to retain the contents.")
+	flag.DurationVar(&opt.cleanupDuration, "delete-after", opt.cleanupDuration, "If namespace exists for longer than this interval, delete the namespace. Set to zero to retain the contents.")
```
Comment (bbguimaraes) [inline cmd/ci-operator/main.go]: Not entirely true now that this depends on a separate controller being present.

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: What would you suggest>

Comment (bbguimaraes) [reply cmd/ci-operator/main.go]: I'd add "(requires openshift/ci-ns-ttl-controller)" or something.

You probably won't care if you're invoking on the command line or your own
cluster, but it's worth documenting.

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: done

File: cmd/ci-operator/main.go [hunk 3]
```diff
@@ -215,7 +212,8 @@ func bindOptions(flag *flag.FlagSet) *options {
 	// the target namespace and cleanup behavior
 	flag.StringVar(&opt.namespace, "namespace", "", "Namespace to create builds into, defaults to build_id from JOB_SPEC. If the string '{id}' is in this value it will be replaced with the build input hash.")
 	flag.StringVar(&opt.baseNamespace, "base-namespace", "stable", "Namespace to read builds from, defaults to stable.")
-	flag.DurationVar(&opt.idleCleanupDuration, "delete-when-idle", opt.idleCleanupDuration, "If no pod is running for longer than this interval, delete the namespace. Set to zero to retain the contents.")
+	flag.DurationVar(&opt.idleCleanupDuration, "delete-when-idle", opt.idleCleanupDuration, "If no pod is running for longer than this interval, delete the namespace. Set to zero to retain the contents. Requires the namespace TTL controller to be deployed.")
+	flag.DurationVar(&opt.cleanupDuration, "delete-after", opt.cleanupDuration, "If namespace exists for longer than this interval, delete the namespace. Set to zero to retain the contents. Requires the namespace TTL controller to be deployed.")
```
Comment (petr-muller) [inline cmd/ci-operator/main.go]: What happens if the TTL controller is *not* deployed? This is annotation-driven, so I assume that the namespace will just live forever, right? Can we detect it and error out or at least warn?

Comment (bbguimaraes) [reply cmd/ci-operator/main.go]: Correct, nothing would remove the namespace.

On the other hand, if it's not deployed, it's most likely a test cluster where
leaking namespaces is not a problem.

Comment (smarterclayton) [reply cmd/ci-operator/main.go]: The first case isn’t true.  It’s trivial easy to run without our infra, and we should keep it that way.  Doesn’t mean we need to support teardown when controller is not present, but don’t apply that philosophy to other parts.

**Comment by smarterclayton** [comment]
This is flaking:

https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/20587/pull-ci-origin-e2e-gcp/2265/

You need to either use patch or retry conflicts multiple times.

File: cmd/ci-operator/main.go (hunk 3 above)
Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: Eh, using `ci-operator` to run your unit tests is trivial. Getting all of the templates, configs and secrets together to run a templatized test on your own _is_ hard, and we are still working through the best way of fixing that. Even Gabe gave up.

**Comment by stevekuznetsov** [comment]
Hmm, will move to `Patch()`.

**Pull request merged**